android/src/main/java/com/liferay/mobile/android/auth/CookieExpirationHandler.java

```diff
@@ -0,0 +1,100 @@
+package com.liferay.mobile.android.auth;
+
+import com.liferay.mobile.android.auth.CookieSignIn.CookieCallback;
+import com.liferay.mobile.android.auth.basic.CookieAuthentication;
+import com.liferay.mobile.android.http.Request;
+import com.liferay.mobile.android.service.Session;
+import com.liferay.mobile.android.service.SessionImpl;
+import com.squareup.okhttp.Authenticator;
+
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CookieExpirationHandler {
+
+    private static final Lock lock = new ReentrantLock();
+    private static final Map<String, Authenticator> authenticators = new HashMap<String, Authenticator>();
+
+    public static Session reloadSessionIfNeeded(final Request request, final CookieCallback callback) throws Exception {
+        final Session session = new SessionImpl(request.getURL(), request.getAuthentication());
+
+        if ((request.getAuthentication() instanceof CookieAuthentication)
+                && ((CookieAuthentication) request.getAuthentication()).shouldHandleExpiration()) {
+
+            CookieAuthentication authentication = (CookieAuthentication) request.getAuthentication();
+
+            lock.lock();
+
+            if (shouldRefreshCookie(authentication.getLastCookieRefresh(),
+                    authentication.getCookieExpirationTime())) {
+
+                Authenticator authenticator = authenticators.get(session.getServer());
+
+                if (callback == null) {
+
+                    try {
+                        Session cookieSession = CookieSignIn.signIn(session, authenticator);
+                        request.setAuthentication(cookieSession.getAuthentication());
+
+                        return cookieSession;
+                    } catch (Exception e) {
+                        throw e;
```

Comment:
why?

Reply:
I don't want to break the current flow, so the exception is handled in the same place than before

Reply:
But you just rethrow it, why not remove that?

Reply:
I need the finally clause to unlock the Lock